src/v/kafka/server/handlers/fetch.cc

```diff
@@ -891,7 +893,17 @@ fetch_handler::handle(request_context rctx, ss::smp_service_group ssg) {
                       [&octx] { return octx.should_stop_fetch(); },
                       [&octx] { return fetch_topic_partitions(octx); });
                 })
-                .then([&octx] { return std::move(octx).send_response(); });
+                .then([&octx] {
+                    // NOTE: Audit call doesn't happen until _after_ the fetch
+                    // is done. This was done for the sake of simplicity and
+                    // because fetch doesn't alter the state of the broker
```

Comment:
I believe fetch may alter state maybe not cluster metadata if thats what your thinking. Question, why was it hard to audit before the fetch? All you need is octx and the request right?

Reply:
> I believe fetch may alter state maybe not cluster metadata if thats what your thinking. 

Any idea what sort of state does get altered?  It won't be consumer offsets, right?

> Question, why was it hard to audit before the fetch? All you need is octx and the request right?

Because of how the fetch planner works.  It 'breaks apart' the fetch request and handles each fetch independently.  So I could call `audit` in each of those "fibers" (for entirely lack of a better word) and stop the fetch right then if the audit fails.  Thoughts?

Reply:
> Any idea what sort of state does get altered? It won't be consumer offsets, right?

No i'm thinking more like reading from disk and filling the batch caches with records, etc

> So I could call audit in each of those "fibers" (for entirely lack of a better word) and stop the fetch right then if the audit fails. Thoughts?

So in all other handlers if we cannot audit then the request is not handled, however in this case the request is handled to completion , then results discarded and error returned. 

I would say lets try to audit before the fetch planner gets the request then , super early in the request handler. I haven't taken a good hard look at it, but if it is something too difficult lets break it into another PR then.

Reply:
> No i'm thinking more like reading from disk and filling the batch caches with records, etc
Innocent question: does it matter?  Or is that more of a performance impact?  My concern with other APIs was that they would alter the cluster state (produce records, delete/create topic, change config, etc).  If this changes something that's transient, I don't think it matters..?

One of the issues is how the response to fetch is built up.  Each of these planners returns a response that gets collated together in the end.  Maybe I can modify that, but it's kind of a complicated handler.  Let me think on it.

To my first point though, if this changes the state of the cluster in a non-transient way, then it should definitely be fixed.